### PR TITLE
bpo-41735: Fix thread locks in zlib module may go wrong in rare case.

### PR DESCRIPTION
**Pull request by ghost**
https://bugs.python.org/issue41735

### CODE REVIEW

**Comment by iritkatriel** [comment]
Can you add a unit test that fails without the fix and passes with it?

**Comment by animalize** [comment]
> 
> 
> Can you add a unit test that fails without the fix and passes with it?

Run this code:
```python
import zlib
import threading

with open('txt.tar', 'rb') as f:
    b = f.read()
    b = b[:100*1024*1024]
print('read', len(b))

cobj = zlib.compressobj()

def f(name, cobj, b):
    print(name)
    dat = cobj.compress(b)
    print(name, len(dat))

thread1 = threading.Thread(target=f, args=('T1', cobj, b))
thread2 = threading.Thread(target=f, args=('T2', cobj, b))
thread1.start()
thread2.start()

print('ok')
```
Before the patch, output:
```
read 104857600
T1
T2
ok
T1 57142315
```
It seems thread2 did not finish the compressing.

After the patch, output:
```
read 104857600
T1
T2
ok
T1 57165411
T2 57166659
```

**Comment by animalize** [comment]
> Can you add a unit test that fails without the fix and passes with it?

Seems a little troublesome. If the data is large, it may be time-consuming. If the data is small, it may not trigger the bug.

Above code can reliably reproduce the bug.

File: Modules/zlibmodule.c [hunk 1]
```diff
@@ -10,10 +10,12 @@
...
-#define ENTER_ZLIB(obj) \
-    Py_BEGIN_ALLOW_THREADS; \
-    PyThread_acquire_lock((obj)->lock, 1); \
-    Py_END_ALLOW_THREADS;
+#define ENTER_ZLIB(obj) do {                      \
+    if (!PyThread_acquire_lock((obj)->lock, 0)) { \
+        Py_BEGIN_ALLOW_THREADS                    \
+        PyThread_acquire_lock((obj)->lock, 1);    \
+        Py_END_ALLOW_THREADS                      \
+    } } while (0)
```
Comment (ambv) [inline Modules/zlibmodule.c:18]: You didn't make this change in the backports. Why not?

Comment (ghost) [reply Modules/zlibmodule.c:18]: This line above the change was added in 3.10.
`zlibstate *state = PyType_GetModuleState(cls);`

So I created PRs separately (3.10 and 3.9-) for code conflict.

The change to `ENTER_ZLIB(obj)` is not a bug fix, maybe only apply it to the master branch?
The code comes from bz2/lzma module without any changes:
https://github.com/python/cpython/blob/v3.10.0a7/Modules/_bz2module.c#L22-L28
https://github.com/python/cpython/blob/v3.10.0a7/Modules/_lzmamodule.c#L18-L24

(`zlib` module was added in 1997, `bz2` module was added in 2002, `lzma` module was added in 2011.)

Comment (ambv) [reply Modules/zlibmodule.c:18]: OK, makes sense.

**Pull request merged**